core/src/calleeFunctions/helpers/uniswapV3.ts

```diff
@@ -15,38 +17,69 @@ const getUniswapV3quoterContract = async function (network: string): Promise<eth
     return new ethers.Contract(UNISWAP_V3_QUOTER_ADDRESS, UNISWAP_V3_QUOTER_ABI, provider);
 };
 
-export const encodeRoute = async function (network: string, collateralSymbols: string[]): Promise<string> {
+export const encodePools = async function (_network: string, pools: Pool[]): Promise<string> {
     const types = [] as string[];
     const values = [] as Array<string | number>;
 
-    for (const collateralSymbol of collateralSymbols) {
+    for (const pool of pools) {
         types.push('address');
-        values.push(await getContractAddressByName(network, collateralSymbol));
+        values.push(pool.addresses[0]);
 
         types.push('uint24');
-        values.push(UNISWAP_FEE);
+        values.push(pool.fee);
     }
-
     types.push('address');
-    values.push(await getContractAddressByName(network, 'MCD_DAI'));
+    values.push(pools[pools.length - 1].addresses[1]);
     return ethers.utils.solidityPack(types, values);
 };
 
-export const convertCollateralToDaiUsingRoute = async function (
+const getRouteAndGasQuote = async (
+    network: string,
+    collateralSymbol: CollateralSymbol,
+    collateralAmount: BigNumber,
+    marketId: string
+) => {
+    const collateral = getCollateralConfigBySymbol(collateralSymbol);
+    const calleeConfig = collateral.exchanges[marketId];
+    const isAutorouted = 'automaticRouter' in calleeConfig;
+    if (calleeConfig?.callee !== 'UniswapV3Callee') {
+        throw new Error(`getCalleeData called with invalid collateral type "${collateral.ilk}"`);
+    }
+    if (isAutorouted) {
+        const { route, quoteGasAdjusted, fees } = await fetchAutoRouteInformation(
+            network,
+            collateralSymbol,
+            collateralAmount.toFixed()
+        );
+        return { route, quoteGasAdjusted, fees };
+    } else {
+        return { route: [collateral.symbol, ...calleeConfig.route], quoteGasAdjusted: undefined, fees: undefined };
+    }
+};
+
+export const convertCollateralToDaiUsingPool = async function (
     network: string,
     collateralSymbol: string,
     marketId: string,
     collateralAmount: BigNumber
 ): Promise<BigNumber> {
     const collateral = getCollateralConfigBySymbol(collateralSymbol);
-    const marketData = collateral.exchanges[marketId];
-    if (marketData?.callee !== 'UniswapV3Callee') {
+    const calleeConfig = collateral.exchanges[marketId];
+    if (calleeConfig?.callee !== 'UniswapV3Callee') {
         throw new Error(`getCalleeData called with invalid collateral type "${collateral.ilk}"`);
     }
     const collateralIntegerAmount = collateralAmount.shiftedBy(collateral.decimals).toFixed(0);
-    const route = encodeRoute(network, [collateral.symbol, ...marketData.route]);
+    const { route, fees } = await getRouteAndGasQuote(network, collateralSymbol, collateralAmount, marketId);
```

Comment:
As suggested above, this function should not generate pools. Moreover, `convertCollateralToDaiUsingPool` inside `uniswapV3` file should _never_ use auto router